### PR TITLE
[ENG-9500] Implement save-cache and restore-cache functions

### PR DESCRIPTION
**Pull request by khamilowicz**
# Why

[ENG-9500](https://linear.app/expo/issue/ENG-9500/caching-for-custom-builds)

User should be able to save and restore files in their custom builds as steps.

This is part 1 of this feature. It provides two new custom build functions: `eas/save-cache` and `eas/restore-cache`. 

```
build:
  name: Foobar
  steps:
    - eas/checkout
    - eas/restore-cache:
      inputs:
        key: cache-key
        paths: node_modules
    - eas/install_node_module
    - eas/save-cache:
        inputs:
          key: cache-key
          paths: node_modules
```

Future PR will allow passing lists of files to `paths` parameter and define dynamic cache-key.

# How

This PR implements `eas/reastore-cache` and `eas/save-cache` functions and lays foundations for future PRs. 
`Cache` type is enhanced with `downloadUrls` attribute which will store mappings between cache keys and signed urls with cached files. Field will be populated by `launcher`. 
`CacheManager` now accepts two parameters: build context (which is either `BuildStepContext` or `BuildContext<Job>`) and `Cache`. Implementation of `CacheManager` using dynamic cache urls will be provided in future PRs. Until this happens, for compatibility, build context is using `any` type. 

# Test Plan

These changes are not meant to provide working functionality, but to provide foundation for new `GCSCacheManager` implementation.

### CODE REVIEW

File: packages/build-tools/src/steps/functions/cache.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        id: 'paths',
+        required: true,
+        allowedValueTypeName: BuildStepInputValueTypeName.JSON,
+      }),
+    ],
+    fn: async (stepsCtx, { inputs }) => {
+      const cacheManager = stepsCtx.global.cacheManager;
+
+      if (!cacheManager) {
+        return;
```
Comment (sjchmiela) [inline packages/build-tools/src/steps/functions/cache.ts:96]: Should we log that something went wrong?

File: packages/steps/src/cacheUtils.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import { bunyan } from '@expo/logger';
+
+export interface Cache {
+  disabled: boolean;
+  clear: boolean;
```
Comment (sjchmiela) [inline packages/steps/src/cacheUtils.ts:5]: What do these do?

File: packages/steps/src/cacheUtils.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  key?: string;
+  cacheDefaultPaths?: boolean;
+  customPaths?: string[];
+  paths: string[];
+  downloadUrls?: Record<string, string>;
```
Comment (sjchmiela) [inline packages/steps/src/cacheUtils.ts:10]:
```suggestion
```

File: packages/steps/src/cacheUtils.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import { bunyan } from '@expo/logger';
+
+export interface Cache {
```
Comment (sjchmiela) [inline packages/steps/src/cacheUtils.ts:3]: Do we need to have two `Cache`s, one here and another one in `eas-build-job`?

File: packages/build-tools/src/steps/functions/cache.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  BuildStepOutput,
+} from '@expo/steps';
+
+function createCacheKey(cacheKey: string, paths: string[]): string {
+  const hash = crypto.createHash('sha256');
+  hash.update(paths.sort().join(''));
+
+  const pathsHash = hash.digest('hex');
+
+  return `${cacheKey}-${pathsHash}`;
```
Comment (sjchmiela) [inline packages/build-tools/src/steps/functions/cache.ts:16]: Should this be the case? That cache key includes paths hash? Or should the cache key be just the key that the user provided? How does Github Actions caching work?

File: packages/steps/src/cacheUtils.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  logger: bunyan;
+  global: CacheableContextCtx;
+  workingdir: string;
+}
+
+export interface CacheManager {
+  // TOOD: Change any to CacheableContext
+  saveCache(ctx: any, cache?: Cache): Promise<boolean | void>;
+  // TOOD: Change any to CacheableContext
+  restoreCache(ctx: any, cache: Cache): Promise<boolean | void>;
```
[the remark this answers is not in the document]
Comment (khamilowicz) [reply packages/steps/src/cacheUtils.ts:29]: `void` is here only to keep compatibility with `GCSCacheManager` in turtle (https://github.com/expo/turtle-v2/blob/9e6fd370c862869fec034d7e1be1074f7d3e3105/src/services/worker/src/CacheManager.ts#L18)

File: packages/steps/src/cacheUtils.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  generateUrls?: boolean;
```
[the remark this answers is not in the document]
Comment (khamilowicz) [reply packages/steps/src/cacheUtils.ts:30]: Yes, it will tell new GCSCacheManager whether to use pregenerated upload/download urls or to generate new ones.

File: packages/steps/src/cacheUtils.ts (hunk 4 above)
Comment (khamilowicz) [reply packages/steps/src/cacheUtils.ts:3]: It is part of the transition process: in order to avoid circular dependencies `Cache` should be defined in `steps`, but to keep compatibility I kept it in both places. It will be consolidated after changes to turtle

File: packages/build-tools/src/steps/functions/cache.ts (hunk 5 above)
Comment (khamilowicz) [reply packages/build-tools/src/steps/functions/cache.ts:16]: GIthub Actions only uses provided key, but the problem would be when user adds new paths to cache without changing key. I can't imagine use case when user doesn't want to automatically update cache file content.

File: packages/steps/src/cacheUtils.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (sjchmiela) [reply packages/steps/src/cacheUtils.ts:29]: My suggestion to make development of this feature easier will be to do as follows:
- make all necessary changes in `eas-build`. They can break API. We shouldn't try to deploy `turtle-v2` with new cache changes (bumped dependencies), but without your adjusting changes (making `GCSCacheManager` conform to new `restoreCache`.
- submit a pull request with upgraded dependencies and updated `turtle-v2` code.

This is the flow I used when developing generic artifacts:
- made a couple breaking changes to `eas-build` (https://github.com/expo/eas-build/pull/306, https://github.com/expo/eas-build/pull/309, https://github.com/expo/eas-build/pull/307)
- updated dependencies and adjusted `turtle-v2` in one PR (https://github.com/expo/turtle-v2/pull/1609).

This removes the complexity of trying to maintain two interfaces at the same time. What do you think, do you foresee any problems?

File: packages/build-tools/src/steps/functions/cache.ts (hunk 5 above)
Comment (sjchmiela) [reply packages/build-tools/src/steps/functions/cache.ts:16]: Yea, makes sense. Also, GitHub _does_ take paths into consideration, see https://github.com/actions/cache?tab=readme-ov-file#cache-version. Maybe there's something else we should take in from what they do.

File: packages/build-tools/src/customBuildContext.ts [hunk 8]
```diff
@@ -56,6 +59,9 @@ export class CustomBuildContext implements ExternalBuildContextProvider {
     this.runtimeApi = {
       uploadArtifacts: (...args) => buildCtx['uploadArtifacts'](...args),
     };
+    this.cacheManager = buildCtx.cacheManager;
```
Comment (sjchmiela) [inline packages/build-tools/src/customBuildContext.ts:62]: We could consider rolling `cacheManager` into `runtimeApi` for simplicity. Like, how many functions may we put into `runtimeApi`? I don't think the list will be too long.

**Comment by khamilowicz** [comment]
@sjchmiela

File: packages/build-tools/src/steps/functions/cache.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+      BuildStepInput.createProvider({
+        id: 'paths',
+        required: true,
+        defaultValue: [],
+        allowedValueTypeName: BuildStepInputValueTypeName.JSON,
+      }),
+    ],
+    outputProviders: [
+      BuildStepOutput.createProvider({
+        id: 'cache_key',
```
Comment (sjchmiela) [inline packages/build-tools/src/steps/functions/cache.ts:31]: nit: I think `key` should be sufficient, same for `paths`.

Or, we could not add it because however users are providing us `key` and `paths`, they should be able to use the same technique anywhere else? This is not a strong opinion.

File: packages/build-tools/src/steps/functions/cache.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+      if (paths.length > 0) {
+        outputs.cache_paths.set(JSON.stringify(paths));
+      }
+
+      if (!cacheManager) {
+        stepsCtx.logger.warn('Cache manager is not available, skipping...');
+        return;
+      }
+
+      const job = stepsCtx.global.staticContext.job as BuildJob;
```
Comment (sjchmiela) [inline packages/build-tools/src/steps/functions/cache.ts:54]: If the function only makes sense for a `BuildJob` let's define argument to this function like https://github.com/expo/eas-build/blob/e15bbd61ccd025b2846d480e0cf454db962b934d/packages/build-tools/src/steps/functions/resolveBuildConfig.ts#L9-L11 so types require a `BuildJob` and we don't need to cast.

However, I think the bigger thing to update here is that I think this doesn't take into consideration updates to https://github.com/expo/turtle-v2/pull/1715, e.g. that cache download URLs are to be requested dynamically from the launcher?

File: packages/steps/build.sh [hunk 11]
```diff
@@ -11,6 +11,7 @@ if [[ "$npm_lifecycle_event" == "prepack" ]]; then
   echo 'Removing "dist_commonjs" and "dist_esm" folders...'
   rm -rf dist_commonjs dist_esm
 fi
+rm -rf dist_commonjs dist_esm
```
Comment (sjchmiela) [inline packages/steps/build.sh:14]: Why did you need to add this? Can we remove it or submit a different pull request explaining the reason for this addition?

File: packages/steps/src/__tests__/BuildConfig-test.ts [hunk 12]
```diff
@@ -33,7 +33,7 @@ describe(readAndValidateBuildConfigAsync, () => {
     const config = await readAndValidateBuildConfigAsync(
       path.join(__dirname, './fixtures/build.yml'),
       {
-        externalFunctionIds: [],
+        externalFunctionIds: ['eas/save-cache'],
```
Comment (sjchmiela) [inline packages/steps/src/__tests__/BuildConfig-test.ts:36]: I can see `eas/*-cache` functions being added as external functions in at least a couple of tests. Why wouldn't they be added as EAS Functions (in `steps/easFunctions.ts`)?

File: packages/steps/package.json [hunk 13]
```diff
@@ -32,6 +32,7 @@
   "license": "BUSL-1.1",
   "devDependencies": {
     "@jest/globals": "^29.7.0",
+    "@types/glob": "^8.1.0",
```
Comment (sjchmiela) [inline packages/steps/package.json:35]: I think we're using `fast-glob` and this is not used/necessary?

File: packages/steps/src/inputFunctions.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+      );
+    })
+  );
+
+  return hash.digest('hex');
+}
+
+export const hashFiles = hashFilesAsync;
+
+export default async function callInputFunctionAsync(
```
Comment (sjchmiela) [inline packages/steps/src/inputFunctions.ts:38]: This (and changes in `template.ts`) are very cool, good job figuring this out! However:

1. Can we not include it in this pull request already? v0 of this pull request may only allow static strings and cache keys and that's ok.
2. Instead of this manual expression parsing and function calling I think we should try to use `jsepEval` on expressions. [This](https://github.com/expo/eas-build/blob/6376bdc493319be4d43e62d1b751fd2161e73cf4/packages/steps/src/BuildStep.ts#L297-L315) is how we're providing functions to `ifCondition`s. It should be simpler and more consistent with other places this way.

File: packages/steps/src/cli/cli.ts [hunk 15]
```diff
@@ -23,7 +23,9 @@ export class CliContextProvider implements ExternalBuildContextProvider {
     public readonly projectSourceDirectory: string,
     public readonly projectTargetDirectory: string,
     public readonly defaultWorkingDirectory: string,
-    public readonly buildLogsDirectory: string
+    public readonly buildLogsDirectory: string,
+    public readonly buildDirectory: string,
+    public readonly projectRootDirectory: string
```
Comment (sjchmiela) [inline packages/steps/src/cli/cli.ts:28]: Oh no, not more directories! haha

What is the reason to add them? Could we use one of the existing directories instead?

If not, can you add comments to those directories explaining what they are? The way we manage directories now is a mess and I find it hard to keep track of what is what.

File: packages/steps/src/cacheUtils.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+  clear: boolean;
+  key?: string;
+  cacheDefaultPaths?: boolean;
+  customPaths?: string[];
+  paths: string[];
+}
+
+export interface DynamicCacheManager {
+  saveCache(ctx: BuildStepContext, cache: Cache): Promise<boolean>;
+  restoreCache(ctx: BuildStepContext, cache: Cache): Promise<boolean>;
```
Comment (sjchmiela) [inline packages/steps/src/cacheUtils.ts:14]: What does `boolean` mean in this context? Could we use `<void>` and throw-and-catch on error?

File: packages/steps/src/cacheUtils.ts (hunk 4 above)
Comment (sjchmiela) [reply packages/steps/src/cacheUtils.ts:3]: Is this still true after recent changes to Turtle? Why would we get a circular dependency if we used `Cache` from `eas-build-job`?

File: packages/steps/src/cacheUtils.ts (hunk 2 above)
Comment (sjchmiela) [reply packages/steps/src/cacheUtils.ts:5]: Unresolving discussion. It remains unclear what does `disabled` or `clear` mean in this context.

File: packages/steps/src/__tests__/fixtures/build.yml [hunk 17]
```diff
@@ -29,3 +29,8 @@ build:
         name: Use non-default shell
         shell: /nib/hsab
         command: echo 123
+    - eas/save-cache:
+        inputs:
+          key: cache-key
+          paths:
+            - src
```
Comment (sjchmiela) [inline packages/steps/src/__tests__/fixtures/build.yml:36]: nit: new line at the end of file. (missing code editor setting on your side?)

```suggestion
            - src

```

File: packages/steps/src/__tests__/BuildWorkflowValidator-test.ts [hunk 18]
```diff
@@ -519,4 +537,29 @@ describe(BuildWorkflowValidator, () => {
       `Custom function module path "/non/existent/module" for function "test" does not exist.`
     );
   });
+
+  test('non-existing  function module', async () => {
```
Comment (sjchmiela) [inline packages/steps/src/__tests__/BuildWorkflowValidator-test.ts:541]:
```suggestion
  test('non-existing function module', async () => {
```

File: packages/steps/src/BuildWorkflowValidator.ts [hunk 19]
```diff
@@ -74,6 +75,28 @@ export class BuildWorkflowValidator {
...
+              }
+              if (fn && fn in inputFunctions) {
+                return null;
+              }
+              return new BuildConfigError(
+                `Input parameter "${currentStepInput.id}" for step "${
+                  currentStep.displayName
+                }" is set to "\${ ${fn}(${args
+                  .map((i) => `"${i}"`)
+                  .join(',')}) }" which is not a valid build-in function name.`
```
Comment (sjchmiela) [inline packages/steps/src/BuildWorkflowValidator.ts:94]:
```suggestion
                  .join(',')}) }" which is not a valid built-in function name.`
```

File: packages/steps/src/BuildStepInput.ts [hunk 20]
```diff
@@ -103,18 +108,65 @@ export class BuildStepInput<
     this.allowedValueTypeName = allowedValueTypeName;
   }
 
-  public get value(): BuildStepInputValueTypeWithRequired<T, R> {
+  public isFunctionCall(): boolean {
+    return this.rawValue?.toString().match(BUILD_STEP_FUNCTION_EXPRESSION_REGEXP) !== null;
```
Comment (sjchmiela) [inline packages/steps/src/BuildStepInput.ts:112]: `rawValue?` means if there is no `rawValue` it will be `undefined` which will be `!== null` and will say it is a function call. Seems unexpected to me?

**Review by sjchmiela — CHANGES_REQUESTED** [review]
Very cool, some questions and suggestions!

File: packages/steps/src/__tests__/BuildConfig-test.ts (hunk 12 above)
Comment (khamilowicz) [reply packages/steps/src/__tests__/BuildConfig-test.ts:36]: `readAndValidateBuildConfigAsync` doesn't know about eas functions, these are passed in `runCustomBuildAsync`(https://github.com/expo/eas-build/blob/afe570422d2c45a7478786e1aeb885a5014d90cf/packages/build-tools/src/builders/custom.ts#L39 and https://github.com/expo/eas-build/blob/afe570422d2c45a7478786e1aeb885a5014d90cf/packages/steps/src/BuildConfigParser.ts#L67).

Comment (sjchmiela) [reply packages/steps/src/__tests__/BuildConfig-test.ts:36]: Hm, maybe let me rephrase my curiosity. Why do we only need to add `eas/save-cache` here and not all other EAS functions? Like, why this function is special and needs to be added here?

Comment (khamilowicz) [reply packages/steps/src/__tests__/BuildConfig-test.ts:36]: Oh, that's because the file that we parse (`./fixtures/build.yml`) only contains steps using either `run` expressions or `eas/save-cache` function. Only functions that are used in template are checked for existence, and `run` is a special case (https://github.com/expo/eas-build/blob/afe570422d2c45a7478786e1aeb885a5014d90cf/packages/steps/src/BuildConfig.ts#L435)

Comment (sjchmiela) [reply packages/steps/src/__tests__/BuildConfig-test.ts:36]: Aaaaah, it's a test file. Yea, sorry, didn't see that.

File: packages/build-tools/src/steps/functions/cache.ts (hunk 10 above)
Comment (sjchmiela) [reply packages/build-tools/src/steps/functions/cache.ts:54]: - call to `restoreCache` depends on `job.cache` existence
- `restoreCache` receives `job.cache`

I thought dynamic caches are completely separate from `job.cache`.
